src/transcoding.ts

```diff
@@ -0,0 +1,318 @@
+
+import {google} from '../protos/http';
+import {RequestType} from './apitypes';
+
+export interface TranscodedRequest {
+  httpMethod: string;
+  url: string;
+  queryString: string;
+  data: string | {};
+}
+
+const httpOptionName = '(google.api.http)';
+
+// The following type is here only to make tests type safe
+type allowedOptions = '(google.api.method_signature)';
+
+const supportedHttpMethods = ['get', 'post', 'put', 'patch', 'delete'];
+
+export type ParsedOptionsType = Array<
+  {
+    [httpOptionName]?: google.api.IHttpRule;
+  } & {
+    [option in allowedOptions]?: {} | string | number;
+  }
+>;
+
+export function getField(
+  request: RequestType,
+  field: string
+): string | number | Array<string | number> | undefined {
+  const parts = field.split('.');
+  let value: RequestType | string | number | undefined = request;
+  for (const part of parts) {
+    if (typeof value !== 'object') {
+      return undefined;
+    }
+    value = (value as RequestType)[part] as RequestType;
+  }
+  if (typeof value === 'object' && !Array.isArray(value) && value !== null) {
+    return undefined;
+  }
+  return value;
+}
+
+export function deepCopy(request: RequestType): RequestType {
+  if (typeof request !== 'object' || request === null) {
+    return request;
+  }
+  const copy = Object.assign({}, request);
+  for (const key in copy) {
+    if (Array.isArray(copy[key])) {
+      copy[key] = (copy[key] as RequestType[]).map(deepCopy);
+    } else if (typeof copy[key] === 'object' && copy[key] !== null) {
+      copy[key] = deepCopy(copy[key] as RequestType);
+    }
+  }
+  return copy;
+}
+
+export function deleteField(request: RequestType, field: string): void {
+  const parts = field.split('.');
+  while (parts.length > 1) {
+    if (typeof request !== 'object') {
+      return;
+    }
+    const part = parts.shift() as string;
+    request = request[part] as RequestType;
+  }
+  const part = parts.shift() as string;
+  if (typeof request !== 'object') {
+    return;
+  }
+  delete request[part];
+}
+
+export function buildQueryStringComponents(
+  request: RequestType,
+  prefix = ''
+): string[] {
+  const resultList = [];
+  for (const key in request) {
+    if (Array.isArray(request[key])) {
+      for (const value of request[key] as RequestType[]) {
+        resultList.push(
+          `${prefix}${encodeWithoutSlashes(key)}=${encodeWithoutSlashes(
+            value.toString()
+          )}`
+        );
+      }
+    } else if (typeof request[key] === 'object' && request[key] !== null) {
+      resultList.push(
+        ...buildQueryStringComponents(request[key] as RequestType, `${key}.`)
+      );
+    } else {
+      resultList.push(
+        `${prefix}${encodeWithoutSlashes(key)}=${encodeWithoutSlashes(
+          request[key].toString()
+        )}`
+      );
+    }
+  }
+  return resultList;
+}
+
+export function encodeWithSlashes(str: string): string {
```

Comment:
Perhaps a comment that explains the encoding happening, is there an RFC we can reference?

Reply:
I have an internal document :(  Maybe makes sense to make it public. For now, I'll just mention the second best document we have.

Reply:
Actually, it's mentioned in line 18 :)